File: json-schema/research/format_materialize_clock/runner.go
// Probe: MATERIALIZE model (B). Parse each validated wire string into Go's
// stdlib typed construct, then re-serialize to the CANONICAL form, and emit
// the bytes. The compare harness checks all materializing languages agree.
//
// Canonical serialization under test:
//   date-time -> UTC-normalized, truncated to ms, "YYYY-MM-DDTHH:MM:SS.mmmZ"
//   date      -> "YYYY-MM-DD"
//   time      -> wall clock, offset DROPPED, ms, "HH:MM:SS.mmm"
//
// go run runner.go corpus.json
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

type row struct {
	ID   string `json:"id"`
	Wire string `json:"wire"`
}
type corpus struct {
	DateTime []row `json:"date-time"`
	Date     []row `json:"date"`
	Time     []row `json:"time"`
}
type out struct {
	ID       string `json:"id"`
	Engine   string `json:"engine"`
	Format   string `json:"format"`
	Canonical string `json:"canonical"` // canonical bytes, or "" if cannot materialize
	Err      string `json:"err"`
}

const engine = "go"

func emit(o out) {
	b, _ := json.Marshal(o)
	fmt.Println(string(b))
}

func main() {
	data, _ := os.ReadFile(os.Args[1])
	var c corpus
	json.Unmarshal(data, &c)

	// date-time : time.Time (RFC3339Nano). Note: Go REJECTS leap :60 and
	// lowercase t/z, so those rows will error -> proves the narrowing needed.
	for _, r := range c.DateTime {
		// parse-path pre-normalization: uppercase the case-insensitive t/z
		// (pinned grammar accepts lowercase; native parser rejects it). Safe
		// because date-time has no other letters (offset is digits only).
		t, err := time.Parse(time.RFC3339Nano, strings.ToUpper(r.Wire))
		if err != nil {
			emit(out{r.ID, engine, "date-time", "", err.Error()})
			continue
		}
		u := t.UTC().Truncate(time.Millisecond)
		emit(out{r.ID, engine, "date-time", u.Format("2006-01-02T15:04:05.000Z07:00"), ""})
	}

	// date : Go has no date type; reuse time.Time via layout.
	for _, r := range c.Date {
		t, err := time.Parse("2006-01-02", r.Wire)
		if err != nil {
			emit(out{r.ID, engine, "date", "", err.Error()})
			continue
		}
		emit(out{r.ID, engine, "date", t.Format("2006-01-02"), ""})
	}

	// time : Go has no time-of-day type. Parse via a time.Time, drop offset,
	// emit wall clock ms. (Layout accepts optional offset.)
	for _, r := range c.Time {
		var t time.Time
		var err error
		// try with offset then without
		w := strings.ToUpper(r.Wire)
		t, err = time.Parse("15:04:05.999999999Z07:00", w)
		if err != nil {
			t, err = time.Parse("15:04:05.999999999", w)
		}
		if err != nil {
			emit(out{r.ID, engine, "time", "", err.Error()})
			continue
		}
		// wall clock components, ms
		ms := t.Nanosecond() / 1e6
		emit(out{r.ID, engine, "time", fmt.Sprintf("%02d:%02d:%02d.%03d", t.Hour(), t.Minute(), t.Second(), ms), ""})
	}
}

File: json-schema/research/format_materialize_clock/runner.py
#!/usr/bin/env python3
"""Probe: MATERIALIZE model (B) in Python via stdlib datetime/date/time.
Parse each validated wire string into the typed construct, re-serialize to the
CANONICAL form, emit the bytes. python3 runner.py corpus.json"""
import json, sys
from datetime import datetime, date, time, timezone, timedelta

ENGINE = "py"

def emit(o):
    print(json.dumps(o))

def canon_datetime(wire):
    # datetime.fromisoformat in 3.11+ handles Z and offsets, but rejects :60
    # and rejects lowercase t/z. Normalize case first to isolate the leap issue.
    # parse-path pre-normalization: uppercase case-insensitive t/z (pinned
    # grammar accepts lowercase; fromisoformat rejects it). Safe: date-time has
    # no other letters (offset is digits only).
    s = wire.upper()
    dt = datetime.fromisoformat(s)  # may raise
    if dt.tzinfo is None:
        raise ValueError("missing offset (naive)")
    dt = dt.astimezone(timezone.utc)
    # truncate to ms
    dt = dt.replace(microsecond=(dt.microsecond // 1000) * 1000)
    ms = dt.microsecond // 1000
    return f"{dt.year:04d}-{dt.month:02d}-{dt.day:02d}T{dt.hour:02d}:{dt.minute:02d}:{dt.second:02d}.{ms:03d}Z"

def canon_date(wire):
    d = date.fromisoformat(wire)
    return f"{d.year:04d}-{d.month:02d}-{d.day:02d}"

def canon_time(wire):
    t = time.fromisoformat(wire.upper())  # handles Z (3.11+) and offsets; rejects :60
    ms = t.microsecond // 1000
    return f"{t.hour:02d}:{t.minute:02d}:{t.second:02d}.{ms:03d}"

def run(rows, fmt, fn):
    for r in rows:
        try:
            emit({"id": r["id"], "engine": ENGINE, "format": fmt,
                  "canonical": fn(r["wire"]), "err": ""})
        except Exception as e:
            emit({"id": r["id"], "engine": ENGINE, "format": fmt,
                  "canonical": "", "err": f"{type(e).__name__}: {e}"})

def main():
    c = json.load(open(sys.argv[1]))
    run(c["date-time"], "date-time", canon_datetime)
    run(c["date"], "date", canon_date)
    run(c["time"], "time", canon_time)

if __name__ == "__main__":
    main()
